### PR TITLE
when logging out, switch to other active user

### PR DESCRIPTION
**Pull request by maxtaco**
- a new Logout engine
- in Login, don't return the global NIST if we're logging in as a different user
- we might want to revisit the whole strategy to logging in as a separate user, since we wind up mutating global state for a failed login, if we're already
logged in as A and we're switching over to B. this is a bigger change, so maybe let's wait on it.

### CODE REVIEW

**Comment by patrickxb** [comment]
How does this affect other calls of G().Logout()?

For example, 

    func (e *loginProvision) cleanup(m libkb.MetaContext) {                              
        if !e.cleanupOnErr {                                     
                return                                                                               
        }                                                  
                                                                                                
        // the best way to cleanup is to logout...
        m.Debug("an error occurred during provisioning, logging out")
        m.G().Logout(m.Ctx())                                                                                             
    }

**Comment by maxtaco** [comment]
That's a good point. It might make sense just to logout in those cases though since I'm not sure login should be calling logout, which then calls login?

**Comment by patrickxb** [comment]
We used Logout in the past to (help) leave the user in a clean state if there was an error.

**Comment by patrickxb** [comment]
Are there any new tests for provisioning when logged in with multiple users?

File: go/engine/login_test.go [hunk 1]
```diff
@@ -177,6 +196,184 @@ func TestUserEmails(t *testing.T) {
...
+
+	// after provisioning, the device keys should be cached
+	assertDeviceKeysCached(tcY)
+
+	// after provisioning, the secret should be stored
+	assertSecretStored(tcY, userX.Username)
+
+}
+
+func TestProvisionAfterSwitchWithWrongUser(t *testing.T) {
```
Comment (patrickxb) [inline go/engine/login_test.go]: 👍

File: go/engine/login_test.go [hunk 2]
```diff
@@ -177,6 +196,184 @@ func TestUserEmails(t *testing.T) {
 	}
 }
 
+func TestProvisionDesktopAfterSwitch(t *testing.T) {
```
Comment (patrickxb) [inline go/engine/login_test.go]: 👍

**Comment by patrickxb** [comment]
LGTM...thanks for the tests.

**Pull request merged**